### DISCUSSION TITLE
Show `node_modules` & Install Fresh from `rails new`?

### DISCUSSION THREAD

**Discussion by dee2k22** [nickjj/docker-rails-example#34] [Q&A]
Love the general idea and workflow of this Docker setup! However, I have a few issues and questions...

1. ~~I prefer having the `node_modules` directory visible/accessible as it allows IDEs to index them and suggest a lot more than without them. Additionally, I often like to dive into the sources when I find documentation to be lacking... How would I go about getting the `node_modules` to appear in my project files? I tried removing the line in the `.yarnrc` file but then I just end up with permission errors when trying to build in WSL2.~~<br>
**Edit:** After re-reading your post, I managed to figure it out... I forgot to edit the `Dockerfile` to represent my desired `node_modules` directory location... `&& mkdir /app/node_modules && chown ruby:ruby -R /app`

2. While cloning a git repository can be convenient, it also means I lack the flexibility of using the runtime flags/args that `rails new` provides. Is there a way to use the Docker files themself as the starting point and run a command to build out a fresh project? Something like `docker run exec -rm web rails new . -T -j esbuild -c tailwind -d postgresql`?

3. Is there a reason you picked `slim-bullseye` over an `alpine` release?

I've been reading your associated blog post and while it's very informative, I'm struggling to understand how to accomplish these two tasks myself and end up with the proper file ownership/permissions inside WSL2...

**Comment by nickjj** [comment]
Hi,

1. It sounds like you got it after your edit. That's how you could do it.

2. You could generate a new Rails project independent of this project and then diff it vs this project. You could spin up a new Ruby based Docker container, set the volume flag, `gem install rails` and now you'd have a new fresh Rails project on your system. This workflow is how this project started. There's no way to incorporate all of these repo's changes into a new Rails project.

3. I've been burned in the past a bunch of times with Alpine. It's custom libc implementation isn't compatible with everything and I've noticed in Python applications the runtime performance was worse when interfacing with Postgres. I'm not sure if the same performance hit is present with Ruby but those 2 things alone are enough to make me happily stick with using Debian.

I use WSL 2 here. File permissions are good to go. Everything is owned by my WSL 2 user (not root). Is it different for you?

**Comment by jon-price-snke** [comment]
Hi @dee2k22, I also need to have access to the `node_modules` folder and I'm wondering how you were able to fix that? I don't quite understand what you did above to make that work. Thanks!

For reference, I'm trying to install `vite_rails` gem. As part of the [installation](https://vite-ruby.netlify.app/guide/#installation-%F0%9F%92%BF) I need to run `bundle exec vite install`, but this fails partway because it needs access to the `package.json` and `node_modules` folder.

**Reply by nickjj** [reply]
What happens if you up everything and run: `./run cmd bundle exec vite install`?

Based on its help menu at https://vite-ruby.netlify.app/guide/#setup-%F0%9F%93%A6 it doesn't seem to do anything other than manipulate config files. Does it have an option to skip running bundle/yarn install automatically in case it does that?

**Reply by jon-price-snke** [reply]
Hi Nick, I tried that before, but here's the steps I took:

1. run `docker-compose up`
2. run `run cmd bundle add vite_rails` (in another terminal tab)
3. run `run cmd bundle exec vite install`

Here's the output:

![bundle exec vite install](https://user-images.githubusercontent.com/63308415/149546366-b179a055-31c4-4903-a980-0d285bfb1224.png)

**Reply by nickjj** [reply]
It's hard to say if that's related to Docker. I never used Vite but it sounds like it's looking for a `config/vite.json` file to exist but it's not there. Are you expected to create it by hand before running the `vite install` command?

**Reply by jon-price-snke** [reply]
No, that `bundle exec vite install` generates all the files. I've actually used it before with docker, but yarn and node_modules were maintained within the repo itself and not in it's custom folder as defined in `.yarnrc`.

That `bundle exec vite install` runs this https://github.com/ElMassimo/vite_ruby/blob/main/vite_ruby/lib/vite_ruby/cli/install.rb and fails on the "install_js_dependencies" [here](https://github.com/ElMassimo/vite_ruby/blob/main/vite_ruby/lib/vite_ruby/cli/install.rb#L21), which is this code:

```
# Internal: Installs vite and vite-plugin-ruby at the project level.
  def install_js_dependencies
    package_json = root.join('package.json')
    write(package_json, '{}') unless package_json.exist?
    deps = js_dependencies.join(' ')
    run_with_capture("#{ npm_install } -D #{ deps }", stdin_data: "\n")
  end
```

I've attached our current (outdated) docker file where this install works fine. I didn't write this, but I've been tasked with updating this app and that's where your awesome setup here comes into play :)

[Dockerfile.txt](https://github.com/nickjj/docker-rails-example/files/7871738/Dockerfile.txt)

**Reply by nickjj** [reply]
The custom node_modules directory should be ok, it's a universal setting.

You could try running this after upping everything: `docker-compose exec js bundle exec vite install`, the Ruby env is available in the assets build stage (image) but this isn't going to build a new image if it happens to yarn install something.

Based on your code snippet, it's sounding like all you'll really have to do is manually add vite's package to your `package.json` and then run `./run yarn:install` and you'll be good to go for installing the front-end bits. Then add the gem to your Gemfile and run `./run bundle:install` to install the Ruby deps, and you're set?

Basically the workflow defined here: https://github.com/nickjj/docker-rails-example#updating-dependencies

**Reply by jon-price-snke** [reply]
Both methods work. After manually adding the yarn packages to package.json and adding the gem to Gemfile, I did the following:

1. run `docker-compose up`
2. run `run yarn:install` (in another terminal tab)
3. run `docker-compose up`
4. run `run bundle:install`
5. run `docker-compose up`
6. run `docker-compose exec js bundle exec vite install` (successfully generates all files needed)
7. run `docker-compose exec js bin/vite dev` to start the vite dev server as noted [here](https://vite-ruby.netlify.app/guide/#running-your-first-example-%F0%9F%8F%83%E2%80%8D%E2%99%82%EF%B8%8F), but this fails as seen below
8. 
![bin vite dev](https://user-images.githubusercontent.com/63308415/149562115-8d60656c-258f-4703-a58c-834d308e2918.png)

So I ran `docker-compose exec js bin/vite info` and it shows me this:

![bin vite info](https://user-images.githubusercontent.com/63308415/149562214-720e3a9c-b307-4a06-ad2a-dd5471a072fb.png)

After digging into the `bin/vite info` code, it's running `npm ls vite vite-plugin-ruby` [here](https://github.com/ElMassimo/vite_ruby/blob/faa548927a87ae5d42065d299dc44c1158b6dae5/vite_ruby/lib/vite_ruby/commands.rb#L109) to check the packages existence. Lol, so I ran `docker-compose exec js npm ls vite vite-plugin-ruby` and it doesn't see them, but if I run `docker-compose exec js yarn list vite vite-plugin-ruby`, it does find them. I'm not sure if this is just something I'm missing on my side or issue with vite rails gem. I do know that I use yarn in the other repo where this all works fine though, so leads me to believe I'm missing something on my side.

I appreciate the help, this is all somewhat new to me.

**Reply by jon-price-snke** [reply]
For reference, here's the github repo that I was playing around with before that uses vite_rails and docker.

https://github.com/ElMassimo/vite_rails_docker_example

**Reply by nickjj** [reply]
In your steps, step 1 and 3 aren't required btw. Both the yarn and bundle install run commands will spin up a fresh container.

In that other repo you linked, Vite is running as a separate container in https://github.com/ElMassimo/vite_rails_docker_example/blob/main/docker-compose.yml.

I don't know what Vite is exactly, is it supposed to be some type of alternative to webpack and esbuild?

**Reply by jon-price-snke** [reply]
Good to know! Still new to docker, so slowly taking it all in.

Ya, I might be over thinking it. What I was trying to do was install and run `vite_rails` within the container itself, but I'm starting to think that I just need to add another vite service as a separate container that can talk to the web container. What would you recommend?

As far as vite goes, it's another frontend tooling option that replaces webpack and various other bundling tools. It's based on esbuild and really fast. https://vitejs.dev/

**Reply by nickjj** [reply]
If it's a separate process that watches files for changes like webpack and esbuild then absolutely I would recommend spinning it up as a separate service. It is its own independent process IMO. This repo follows that pattern too. Back when it used Webpacker, that was a separate process and now with esbuild that is its own process. It even splits tailwind into its own process too and both esbuild and tailwind share the same Docker image.

**Reply by jon-price-snke** [reply]
Thanks, I appreciate the information and the help! I'm going to play around with it and see if I can get it up and running as a separate container. Might have more questions for you :)
